rfcs/0024-versioning-and-backwards-compatability.md

```diff
@@ -0,0 +1,349 @@
+## Summary
+[summary]: #summary
+
+This is a proposal to ensure that we maximise backwards compatability across
+all of our communication protocols.
+
+This follows on from [RFC 0015: Module Versioning](0015-module-versioning.md),
+aiming to keep the same underlying module structures but better enforcing their
+registration.
+
+## Motivation
+
+[motivation]: #motivation
+
+Our current system for versioning is split into type versioning and type
+registration, both of which need to happen to enforce backwards compatability.
+The registration stage is not enforced and so may be easily missed.
+
+Further, the registration stage requires some extra boilerplate code that could
+easily be generated by an OCaml ppx. Using a ppx could also ensure that this
+code is always present when it should be.
+
+Finally, there are complexities around functors and types with parameters that
+were not addressed in the original proposal.
+
+## Detailed design
+
+[detailed-design]: #detailed-design
+
+### Module structure
+
+The proposed structure closely mirrors that of the original [RFC
+0015](0015-module-versioning.md):
+
+```ocaml
+module Module_name = struct
+  module Stable = struct
+    [%%versioned]
+
+    module V2 = struct
+      type t = ...
+    end
+
+    module Latest = V2
+
+    module V1 = struct
+      type t = ...
+
+      let to_latest t = ...
+    end
+  end
+
+  type t = Stable.Latest.t
+end
+```
+
+The ppx that triggers on `versioned` should then convert this module structure
+to be compatible with (a slightly tweaked version of) our existing registration
+system:
+
+```ocaml
+module Module_name = struct
```

Comment:
in some cases, we use `Binable.of_string` rather than `deriving bin_io`. Maybe special-case that.

Reply:
Actually, it was `Binable.Of_binable` I was thinking of. That functor shadows existing `bin_io` functions. See issue #3388. The call to this functor can just be added to a module definition, and will work as long as it follows the registration code. So no special case needed, I think.